core/src/services/analytics/sboms/service.rs

```diff
@@ -0,0 +1,359 @@
+                "pipeline": [
+                    {
+                        "$project": {
+                            "_id": 0,
+                            "severity": 1,
+                            "cve": 1,
+                            "description": 1,
+                            "cvss": 1,
+                            "cwes": 1,
+                            "remediation": 1
+                        }
+                    }
+                ]
+            }
+        })
+    )
+}
+
+fn get_group_purls_stage() -> Stage {
+    Stage::new(
+        json!({
+            "$group": {
+                "_id": "",
+                "purls": {
+                  "$push": {
+                    "purl": "$purl"}
+                    }
+                }
+            }
+        )
+    )
+}
+
+fn get_group_stage() -> Stage {
+    Stage::new(
+        json!({
+            "$group": {
+                "_id": "$_id",
+                "name": {
+                    "$first": "$name"
+                },
+                "packageManager": {
+                    "$first": "$packageManager"
+                },
+                "purl": {
+                    "$first": "$purl"
+                },
+                "provider": {
+                    "$first": "$provider"
+                },
+                "version": {
+                    "$first": "$version"
+                },
+                "created": {
+                    "$first": {
+                        "$toDate": {
+                            "$multiply": [ "$timestamp", 1000 ]
+                        }
+                    }
+                },
+                "report": {
+                    "$push": {
+                        "name": "$report.name",
+                        "version": "$report.version",
+                        "purl": "$report.purl",
+                        "packageManager": "$report.packageManager",
+                        "enrichments": [
+                            "$report.snyk_enrichment"
+                        ]
+                    }
+                }
+            }
+        })
+    )
+}
+
+fn get_unset_stage() -> Stage {
+    Stage::new(
+        json!({
+            "$unset": "_id"
+        })
+    )
+}
+
+/// Service to create and run analytics on DocumentDB
+#[derive(Clone)]
+pub struct AnalyticService {
+    pub(crate) store: Arc<MongoStore>,
+    pub(crate) storage: Arc<dyn StorageProvider>,
+    pub(crate) analytic: Analytic,
+}
+
+impl AnalyticService {
+
+    /// Creates a new AnalyticService
+    pub fn new(store: Arc<MongoStore>, storage: Arc<dyn StorageProvider>) -> Self {
+
+        let analytic = Analytic::new(store.clone());
+
+        AnalyticService {
+            store,
+            storage,
+            analytic,
+        }
+    }
+
+    /// Queries MongoDB to get all of the purls for the primary SBOMs
+    pub async fn get_primary_purls(&mut self) -> Result<Option<Vec<String>>, PlatformError> {
+
+        self.analytic.add_stage(
```

Comment:
Yeah, figured a few things out from your earlier comment.  I ended up using a Mutex in the Analytic service. If that doesn't affect performance too terribly (as locking threads can do), I'm happy with that route.

Reply:
I'm leaving this comment open to see if we should refactor the code.

Reply:
Also, keep in mind that add_stage() wasn't where this ended.  It was the pipeline underneath that I needed to push data into that needed to be mutable.  The Mutex seems to have solved the issue, however I didn't implement your suggested fix, so I'm leaving this conversation open in case using a Mutex isn't acceptable and we need to refactor again to go with what you suggested.